### ISSUE TITLE
Cannot alias :reload in a chained command

### ISSUE THREAD

**Issue by RomanSC** [The-Compiler/qutebrowser#1850]
Sorry for the cryptic bug name, I'm not sure if you can't alias multiple commands into one, or just :reload.

This alias should disable javascript and reload the page when the :js is executed. Instead attempting to alias with the below command:

set aliases js "set content allow-javascript! ;; reload" 

Returns: 

reload": no such command

Version info: 

qutebrowser v0.8.2
Git commit: 

CPython: 3.5.2
Qt: 5.7.0 (compiled 5.6.0)
PyQt: 5.6

sip: 4.18
colorama: no
pypeg2: 2.15
jinja2: 2.8
pygments: 2.1.3
yaml: 3.11
cssutils: no
typing: yes
PyQt5.QtWebEngineWidgets: no
pdf.js: no
Webkit: 538.1
Harfbuzz: old
SSL: OpenSSL 1.0.2h  3 May 2016

Platform: Linux-4.7.0-1-ARCH-x86_64-with-arch-Arch-Linux, 64bit
Desktop: /usr/share/xsessions/i3-with-shmlog
Frozen: False
Imported from /usr/lib/python3.5/site-packages/qutebrowser
OS Version: 

--- /etc/lsb-release ---
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"

--- /etc/arch-release ---
Arch Linux release

--- /etc/os-release ---
NAME="Arch Linux"
ID=arch
PRETTY_NAME="Arch Linux"

**Comment by The-Compiler** [comment]
The easy way out of this would be for `:set` to simply not split commands with `;;` - however I'm starting to wonder if we should *never* split when `;;` is used in the commandline directly, only when it's used via an alias or keybinding...

**Comment by blyxxyz** [comment]
I wouldn't expect `;;` to split at all if it's quoted.

**Comment by The-Compiler** [comment]
@blyxxyz That's a good point - no idea off-hand how easy that'd be to implement though.

**Comment by ryanjmortenson** [comment]
I have got around this issue by temporarily replacing the quoted `;;` with another set of characters before the split then returning them to their original state before passing into the `sub_text` parse.

**Comment by The-Compiler** [comment]
#1895 also has some discussion on this FWIW.

**Comment by lahwaacz** [comment]
I think using quotes to escape `;;` (and potentially also `{variables}` in #1861) is practically impossible as long as [simple_split](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/misc/split.py#L181-L211), which does not understand quotes, is in use for some commands (e.g. `:bind` and `:set-cmd-text`). It is very messy already that some commands handle quotes literally and some don't, some interpret `;;` literally, some don't expand `{url}`. IMO the [maxsplit](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/commands/command.py#L70-L71), [no_cmd_split](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/commands/command.py#L79) and [no_replace_variables](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/commands/command.py#L82) command attributes should be replaced with proper escaping rules. In theory, quoting should solve everything - instead of `:bind key command --with flag` you'd write `:bind key "command --with flag"`, not a big deal.

**Comment by The-Compiler** [comment]
I don't think quoting alone is sufficient, we already have that. But I wouldn't want to use `:open "some search"` all the time, and when you want to mix `:bind` and `:set-cmd-text` with something with spaces you might not want three levels of quoting either :wink:

**Comment by lahwaacz** [comment]
With [argparse](https://docs.python.org/3/library/argparse.html#example) you can use `nargs='+'` to accumulate all remaining positionals, I'd expect `:open` to behave that way if there was no `simple_split`. The problem would be passing quotes to the search engine.

For combining `:bind` and `:set-cmd-text` there are only two levels, so you can alternate single and double quotes. Both commands could behave as above, so the quotes wouldn't be necessary unless you need to escape `;;` or something else.

**Comment by The-Compiler** [comment]
> The problem would be passing quotes to the search engine.

That, and not handling multiple spaces between arguments correctly (which might be less a problem for `:open` than for `:search`, but still is a problem).

Since escapes don't work in single quotes (at least with how splits are done currently, like in a shell), when you'd want to bind something to `:open` with multiple args and want that to work correctly, you'd need to do `:bind x "set-cmd-text \":open \\"foo bar\\"\""`, right?

**Comment by lahwaacz** [comment]
Not sure what you mean by escapes not working in single quotes, but this seems to do the right thing:
```python
>>> from qutebrowser.misc.split import *
>>> text = r"""
... :bind x "set-cmd-text ':open \"foo bar\"'"
... """
>>> split(text)
['\n:bind', 'x', 'set-cmd-text \':open "foo bar"\'\n']
```
`shlex.split` gives the same output (except for the `\n` at the ends).

**Comment by lahwaacz** [comment]
But the point of https://github.com/The-Compiler/qutebrowser/issues/1850#issuecomment-246273921 is that if there are commands that interpret quotes literally and we use quotes to escape `;;` then you can't escape `;;` in these commands.

**Comment by Kingdread** [comment]
I agree with lahwaacz and I think that the currenzlt way of doing things with all those "special behavior" commands will get messy (with stuff like this issue or the variable replacement). A proper quoting system should be able to deal with many special needs. Some thoughts:

* There should be a quoting that still replaces variables, for stuff like `:yank "{url} - {title}" ` where you want multiple variables in one argument.
* We probably don't need the weird `Hello" World"` "interwoven" quoting.
* We could take inspiration from Python's raw strings, or Rust's: You can do stuff like `r###"I can use " and # freely here"###` where you need the right `"` + correct number of `#` combination to close the string. While allowing maximum flexibility, it comes at the cost of verbosity. We could change that syntax to something lightweight like `r3"foobar"3` (or maybe `."` as regex).
* We could use a quoting style with "real counterpart", like `{` and `}` or `(` and `)` (possibly with a modifier like `%(this is a string)`). I think Perl and Ruby have something similar. This could allow us to best strings without escaping, like `%(level 1 %(level 2))`, as the parser knows how many levels are open.

I can add documentation links later, but Icm currently on the phone in the train :wink:

**Comment by lahwaacz** [comment]
As for the literal interpretation of quotes, maybe the tokenizer could always use the quotes for quotation/escaping, but not strip the quotes. Instead, the commands themselves would call something like `strip_quotes()` on each argument as necessary. Or the tokenizer could yield pairs of both variants and the commands would choose whichever they like.

**Comment by The-Compiler** [comment]
From https://github.com/qutebrowser/qutebrowser/pull/4081#issuecomment-407678514:

Another possibility: Something like herbstluftwm's `chain` command (see its [manpage](http://herbstluftwm.org/herbstluftwm.html)), which would allow you to do:

```
:chain ;; jseval foo ;; jseval bar
```

but it always splits, and you can chose your own separator as the first argument.
